[AMDGPU] IGLP: Fix static variables

ro-i opened the pull request:
Replace global / class-level static variables with instance members and guarantee thread-safety.

@arsenm as discussed

I tried to keep the changes as non-invasive as possible and without functional change.

ro-i commented on llvm/lib/Target/AMDGPU/AMDGPUIGroupLP.cpp:
```diff
@@ -1353,17 +1344,25 @@ bool MFMAExpInterleaveOpt::analyzeDAG(const SIInstrInfo *TII) {
+      //   (with potential arithmetic between exp and cvt)
+      //   see
+      //   https://github.com/llvm/llvm-project/pull/80370#discussion_r1483660378
+      if (SU.Succs.size() >= 7 && any_of(SU.Succs, IsFMACDataDep))
```
I would like to add at this point that this heuristic seems highly problematic to me. It works on all test cases, but I have no documentation which states that this holds in general (same for the previous heuristic). It's just what I observed and a necessary change to be able to handle with the post-ra situation.

ro-i replied:
So please let me know if there is a better way to detect the additional EXP (also in post-ra)

ro-i replied:
I forgot to mention. The problem basically may be the following (using llvm.amdgcn.iglp.opt.exp.small.mir as an example).
Take a look at the following snippet from the initial MIR:
```
  undef %247.sub0:vreg_64_align2 = afn nofpexcept V_EXP_F32_e32 %246:vgpr_32, implicit $mode, implicit $exec
[...]
  %0:vgpr_32 = contract nofpexcept V_FMAC_F32_e32 %1:vgpr_32, %247.sub0:vreg_64_align2, %0:vgpr_32, implicit $mode, implicit $exec
  %259.sub0_sub1:vreg_512_align2 = contract nofpexcept V_PK_MUL_F32 8, %259.sub0_sub1:vreg_512_align2, 0, %247:vreg_64_align2, 0, 0, 0, 0, 0, implicit $mode, implicit $exec
[... a lot more V_PK_MULs ... ]
```
If I'm not mistaken, this basically means that only the lower 32bit of `%247` are actually defined. Now take a look at the following snippet taken from directly before postmisched runs:
```
  renamable $vgpr16 = afn nofpexcept V_MUL_F32_e32 1069066811, killed $vgpr16, implicit $mode, implicit $exec
  renamable $vgpr17 = afn nofpexcept V_MUL_F32_e32 1069066811, killed $vgpr17, implicit $mode, implicit $exec
[... vgpr16 and vgpr17 are used and vgpr16 is also reassigned ... ]
  renamable $vgpr16 = afn nofpexcept V_EXP_F32_e32 killed $vgpr24, implicit $mode, implicit $exec
[...]
  renamable $vgpr48_vgpr49 = contract nofpexcept V_PK_MUL_F32 8, $vgpr48_vgpr49, 0, $vgpr16_vgpr17, 0, 0, 0, 0, 0, implicit $mode, implicit $exec
[... a lot more V_PK_MULs ... ]
[...]
 dead renamable $vgpr2 = contract nofpexcept V_FMAC_F32_e32 killed $vgpr4, killed $vgpr16, killed $vgpr2(tied-def 0), implicit $mode, implicit $exec
```
I think that vgpr17 is only "used" by V_PK_MUL_F32 because this instruction requires two dwords as parameter. It's value is not actually defined, which is why it's not detected to have more then 7 successors in the runs before RA.

ro-i replied:
That's why there is only one SU with more than 7 Succs before RA but two in post-RA

arsenm replied:
If we have a partially undef vector feeding a trivially scalarizable operation, that's an upstream optimization failure. It's also fine to not detect the undefined half as a meaningful dependency

ro-i replied:
> It's also fine to not detect the undefined half as a meaningful dependency

see https://github.com/llvm/llvm-project/pull/140255

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPUIGroupLP.cpp:
```diff
@@ -229,13 +229,13 @@ class SchedGroup {
              ScheduleDAGInstrs *DAG, const SIInstrInfo *TII)
       : SGMask(SGMask), MaxSize(MaxSize), DAG(DAG), TII(TII) {
-    SGID = NumSchedGroups++;
+    SGID = __atomic_fetch_add(&NumSchedGroups, 1, __ATOMIC_SEQ_CST);
```
Nothing should be atomic here

ro-i replied:
Why not? (Maybe I misunderstood something)

arsenm replied:
There is no multithread compilation or shared state. The thread safety risk is from different LLVMContexts in different threads, not within the single compilation context. Multiple sequential compiles in the same process are also broken by static variables

ro-i replied:
alright, thanks

frederik-h commented on llvm/test/CodeGen/AMDGPU/iglp-cache-serialization.mir:
```diff
@@ -0,0 +1,85 @@
+  bb.1:
+    liveins: $vgpr0, $vgpr1
+
+    IGLP_OPT 1
```
Add test for IGLP_OPT 2.

frederik-h reviewed (commented):
I am not sure that the performance improvements achievable by the caching are worth the added code complexity. In fact, I did a quick experiment and I changed the baseline `MFMASmallGemmSingleWaveOpt::applyIGLPStrategy` implementation to always recompute the analysis data and compared the time spent in the scheduler of that build with the time using a build of this PR. For this I used different IR files from a build of project that uses "IGLP_OPT 1" (composable_kernel). Recomputing was consistently faster for me. Perhaps that's different for the other strategy or perhaps I made a mistake in my ad hoc benchmarking, but in any case I would suggest that you also try to benchmark if you haven't already. I think @jrbyrnes has originally implemented those strategies. @jrbyrnes Do you think the caching is crucial? @ro-i If you do not see a significant performance advantage obtained by the caching, perhaps you could fix the correctness issue first (recomputing the analysis data) and then try to improve the performance?

ro-i committed:
[AMDGPU] IGLP: Fix static variables

Replace global / class-level static variables with proper caches and
guarantee thread-safety.

Assisted-by: claude-opus-4.6

ro-i commented:
Since there seems to be a bit more traffic on IGLP stuff recently (and I would really love to get this PR out of my list), I updated the whole thing and also addressed @frederik-h's comments.
I'm closing the alternativ PR since the original author opposed (although the fact that the scheduling was different due to recomputation may not imply that it's wrong.)

ro-i commented:
The whole IGLP stuff seems a little messy, but with fixed static variables, it's at least not wrong anymore

shiltian commented:
What is the motivation of making this thread safe? I'm not sure whether LLVM is developed in a thread safe way.

ro-i commented:
There is an internal user that reported that he encountered a thread safety issue for a big gemm kernel.
Also, it's not just about the thread safety. The static variables weren't even properly reset before, which also led to issues (see the test in iglp.opt.ll that I added and that would crash without the patch).

shiltian commented:
> The static variables weren't even properly reset before

This needs to be fixed, but thread safety isn't something we need to worry about in LLVM. LLVM has a lot of global state that can't really work in a multithreaded environment without proper protection.

hidekisaito commented on llvm/lib/Target/AMDGPU/SIMachineFunctionInfo.h:
```diff
@@ -552,6 +555,27 @@ class SIMachineFunctionInfo final : public AMDGPUMachineFunction,
+    bool HasCvt = false;
+    bool HasChainBetweenCvt = false;
+    bool AnalysisResult = false;
+  };
```
Is the notion of operand bundle or similar applicable at Machine IR level? If so, attaching such info in the form of operand bundles to IGLP_OPT call site appears most natural to me.

arsenm replied:
You can embed whatever you want in the pseudo instruction, it doesn't need to be a bundle or anything

jrbyrnes commented:
> > The static variables weren't even properly reset before
> 
> This needs to be fixed, but thread safety isn't something we need to worry about in LLVM. LLVM has a lot of global state that can't really work in a multithreaded environment without proper protection.

+1 That the counters should be reset if we are in `SchedulingPhase::Initial` and any asserts on StaticVariable == 0 should be fixed for the case where we have multiple functions using the iglp.opt per TU (though this does feel more of a manufactured error than a true user error).

arsenm commented:
> This needs to be fixed, but thread safety isn't something we need to worry about in LLVM. LLVM has a lot of global state that can't really work in a multithreaded environment without proper protection.

This is not true, and it is something we need to worry about. You can have different LLVMContexts in different threads for parallel compilers.

Additionally not all use cases are single compile per process. Any runtime compile environment will be reading dirty state from the last compile, which is fully sequential

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPUIGroupLP.cpp:
```diff
@@ -893,30 +903,27 @@ bool MFMASmallGemmOpt::applyIGLPStrategy(
-  // Whether or not there are instructions between the TRANS instruction and
-  // V_CVT
-  static bool HasChainBetweenCvt;
+  // The heuristics for the pipeline:
```
These comments belong with the new struct definition?

ro-i replied:
(I deliberately put them here because that's the location where you would want to read them)

arsenm commented on llvm/lib/Target/AMDGPU/SIMachineFunctionInfo.h:
```diff
@@ -614,7 +679,11 @@ class SIMachineFunctionInfo final : public AMDGPUMachineFunctionInfo,
   IndexedMap<uint32_t, VGPRBlock2IndexFunctor> MaskForVGPRBlockOps;
 
-private:
+  DenseMap<const MachineInstr *, MFMASmallGemmSingleWaveCache>
```
This needs documentation. It's also kind of terrible that there's an instruction pointer here

ro-i replied:
the cache has only meaning in combination with the corresponding IGLP instruction

arsenm commented on llvm/lib/Target/AMDGPU/SIMachineFunctionInfo.h:
```diff
@@ -259,6 +279,45 @@ template <> struct MappingTraits<SIMode> {
+
+template <> struct MappingTraits<MFMASmallGemmSingleWaveCacheEntry> {
+  static void mapping(IO &YamlIO, MFMASmallGemmSingleWaveCacheEntry &E) {
+    YamlIO.mapRequired("mbb", E.MBBNum);
```
Why is the block reference needed? Isn't that just the block the IGLP_OT instruction is in?

ro-i replied:
exactly. And I wanted to avoid putting the instruction pointer there, too

ro-i replied:
that's why I opted for mbb + index within mbb as a workaround

arsenm commented on llvm/lib/Target/AMDGPU/SIMachineFunctionInfo.h:
```diff
@@ -86,6 +88,24 @@ class AMDGPUGWSResourcePseudoSourceValue final : public AMDGPUPseudoSourceValue
 };
 
+struct MFMASmallGemmSingleWaveCache {
+  unsigned DSWCount = 0;
```
I'm not sure what "DSW" is supposed to stand for

shiltian commented:
> This is not true, and it is something we need to worry about. You can have different LLVMContexts in different threads for parallel compilers.

I'd argue the `LLVMContext` per thread is the maximum that we can promise at LLVM level but IIRC even with `LLVMContext` per thread, it is not really thread safe. I definitely ran into some weird issue when I was working on JIT compilation for OpenMP back a few years ago, and I did have one context per thread.

However, the real problem that this PR is trying to resolve is not really related to multi-threading. The static counters should definitely be reset properly.

ro-i committed:
fix comments and make output order deterministic

jrbyrnes commented on llvm/lib/Target/AMDGPU/AMDGPUIGroupLP.cpp:
```diff
@@ -1354,11 +1348,17 @@ bool MFMAExpInterleaveOpt::analyzeDAG(const SIInstrInfo *TII) {
+      //   fma_f32 -> exp_f32 -> cvt_f16_f32 -> v_pack_b32_f16 -> mfma_.._f16
+      //   (with potential arithmetic between exp and cvt)
+      //   see
+      //   https://github.com/llvm/llvm-project/pull/80370#discussion_r1483660378
```
Not following this comment -- what exactly needs clarification?

The convo linked here is regarding why we were limiting to just v_fma_f32 - which seems to be independent from this part of the code?

shiltian replied:
A side note: not a good practice to include this kind of URL in the code.

ro-i replied:
wanted to spare future people the lookup

jrbyrnes commented:
> > This needs to be fixed, but thread safety isn't something we need to worry about in LLVM. LLVM has a lot of global state that can't really work in a multithreaded environment without proper protection.
> 
> This is not true, and it is something we need to worry about. You can have different LLVMContexts in different threads for parallel compilers.
> 
> Additionally not all use cases are single compile per process. Any runtime compile environment will be reading dirty state from the last compile, which is fully sequential

This is a very niche and targeted optimization, with a very specific use case.  AFAIK, the users of this builtin are not doing any runtime / parallel compilation, and personally I don't think we should be spending much time worrying about those use cases. That was sort of the original intent of this when I left it as an assertion failure if we ran into unexpected compilation flow, though it would've been better encoded as a report_fatal_error -- this is also why I'm a bit curious how we're even running into this.

Anyways, the reason why we needed to fail for unsupported compilation flows is because iglp framework is lacking a caching mechanism. Our mutations are dependent upon the structure of the software pipeline as inferred by the RAW dependencies of the DAG - but the dependencies change in PostRA scheduling, so we need to either have complex dependency analysis PostRA or reuse our previous analysis.

I do think it would be beneficial to add a generalized caching system for all iglp.opts to use - but, if that is not the direction this is going, I'd just as soon say use report_fatal_error on that static variables than spend a bunch of effort to get iglp.opt 2 to support compilation flows it doesn't really need to.

arsenm commented:
> I do think it would be beneficial to add a generalized caching system for all iglp.opts to use - but, if that is not the direction this is going, I'd just as soon say use report_fatal_error on that static variables than spend a bunch of effort to get iglp.opt 2 to support compilation flows it doesn't really need to.

IMO this gets fixed or we remove all of this. It needs to be reimplemented or deleted. You can't make just let things be broken on arbitrary observed user patterns

ro-i committed:
remove FIXME

jrbyrnes commented:
> > I do think it would be beneficial to add a generalized caching system for all iglp.opts to use - but, if that is not the direction this is going, I'd just as soon say use report_fatal_error on that static variables than spend a bunch of effort to get iglp.opt 2 to support compilation flows it doesn't really need to.
> 
> IMO this gets fixed or we remove all of this. It needs to be reimplemented or deleted. You can't make just let things be broken on arbitrary observed user patterns

Surely this wouldn't be the first time that "using the builtins / compiler in an unsupported way is an error" has occurred. But I get it, these iglp.opts in general leave a bad taste in the mouth. 

I agree that we can do better for this specific part, and, if we're not going to, then we should just drop support for it. But we can't just delete it today, we need a deprecation plan since we have users who are depending on this for optimized code. The deprecation plan likely involves replacing it with some of the new scheduling features we have coming up.

ro-i commented:
ping
